Refactor workflow

peterk87 opened the pull request:
General refactoring and clean-up of workflow:

## Changes

- refactored Python scripts for clarity using type hinting where possible
  - using `shiptv=0.4.1` package and containers for most processes since Python dependencies should already be installed (rich, typer, biopython, pandas, numpy)
  - typer for CLI argument parsing
  - added logging with Rich
  - for reading metadata into Pandas DataFrames, setting first column as index (`index_col=0`)
  - for merging metadata dataframes, using `pd.concat` on index instead of `pd.merge`
  - renamed Python scripts to match Nextflow process names
- `bin/filter_gisaid.py`:
  - added concatenation of user sequences and filtered GISAID sequences to reduce duplication of data (no longer need `CAT_SEQUENCES` process)
  - added `--max-gisaid-seqs` to limit number of GISAID sequences to pull from large lineages
  - changed stats output to JSON for easier programmatic use
- reorganized results output directory structure
- moved all process modules from `modules/nf-core/software/` to `modules/local/`; `modules/nf-core/software/` should be reserved for *unmodified* [nf-core tools modules](https://github.com/nf-core/modules) only
- renamed processes to match program name where possible (e.g. `IQTREE_PHYLOGENETICTREE` to `IQTREE`)
- `nextflow_schema.json` descriptions and organization
- `aa_substitution_change` to `aa_mutation_matrix`
- process output files explicitly defined where possible (`path '*.tsv'` to `path 'metadata.filtered.tsv'`)
- updated docs in `README.md`, `docs/usage.md`; `output.md` needs to be updated still
- `bin/merge_metadata.py`: combining Pangolin results and GISAID metadata into the same fields
- Nextalign now using more cores (`--jobs ${task.cpus}`) and output files now explicit
removed unused process nw_reroot.nf
- `params.visualize_gisaid_metadata` to `params.select_gisaid_metadata`; fixed issue with `--select_gisaid_metadata` workflow param due to default value set as "''"


## Added

- software version scraping and output to tab-delimited table `pipeline_info/software_versions.tsv`
- `--max_gisaid_filtered_seqs` to limit the number of GISAID sequences to pull from large lineages when running `bin/filter_gisaid.py` (default: 100000)
- workflow params to skip Nextclade (`--skip_nextclade`) and R SNP tree processes (`--skip_snp_tree`)

## Updates

- shiptv version 0.4.1

## Removed

- process `CAT_SEQUENCES` and `modules/nf-core/software/cat/`
- `emit: <name>` for processes with only one output channel
- unused/unnecessary files from repo
  - GitHub workflows
  - email assets (workflow not configured for sending emails)
  - `conf/igenomes.config`
- `bin/prune_down_tree_ete3.py`; unused
- `modules/local/nw_reroot.nf`; unused
- `Dockerfile`; unused and not needed

## CI

- removed `nf-core lint` GitHub Action workflow since this workflow is not going to be added to nf-core at this time
- added artifact uploading. `shiptv.html` on CI test pass; `.nextflow.log` and `iqtree-*.log` on fail.

## IQ-TREE issues

IQ-TREE v2.1.2 seems to have sporadic issues with CI; others have seen the same issue:

- [BUG: IQTree 2.1.12 fails to run on input with ERROR: Tree taxa and alignment sequence do not match (see above)](https://groups.google.com/g/iqtree/c/ZxrAty8br0E)
- [iqtree/iqtree2 issue 11](https://github.com/iqtree/iqtree2/issues/11)

I'm working on updating IQ-TREE to version 2.1.4-beta in Bioconda.

## PR checklist

- [X] This comment contains a description of changes (with reason).
- [X] If you've fixed a bug or added code that should be tested, add tests!
  - [ ] If you've added a new tool - add to the software_versions process and a regex to `scrape_software_versions.py`
  - [ ] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/scovtree/tree/master/.github/CONTRIBUTING.md)
- [X] Ensure the test suite passes (`nextflow run . -profile test,docker`).
- [X] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

nhhaidee committed:
add support for multiple lineages

peterk87 committed:
refactor workflow

peterk87 committed:
Fix IQ-TREE issue "ERROR: Tree taxa and alignment sequence do not match"; add `--keep-ident` to iqtree command (https://github.com/Cibiv/IQ-TREE/issues/196)

peterk87 committed:
Fix IQ-TREE CI issue; use default initial tree instead of NJ-R
Move IQ-TREE args to modules.config

peterk87 committed:
Fix IQ-TREE CI GISAID test issue

nhhaidee commented:
@peterk87 For IQTree issue, there is no problem when I ran tests locally, still don't know why it failed intermittently with CI

peterk87 commented:
> @peterk87 For IQTree issue, there is no problem when I ran tests locally, still don't know why it failed intermittently with CI

@nhhaidee It may be an issue with `-t NJ-R` producing a tree with certain taxa dropped. Might be an issue with running it on Ubuntu or specific hardware. If the default `fast initial parsimony tree by random order stepwise addition` is used, then CI passes.

nhhaidee commented:
OK.
FYI: for nf-core-lint failed with files_unchanged, please update these files .nf-core-lint.yaml. I am running refactor workflow with my own data

peterk87 committed:
general cleanup and fixes
reorganize results output
added workflow params descriptions to usage.md
removed unnecessary GitHub Actions CI tests, linting and docker image building workflows
removed unused email assets
removed Dockerfile

nhhaidee commented:
`/bin/filter_msa.py`:

`df = pd.read_table(input_metadata, index_col=0)`: dataframe of input_metadata does not contain strains of input sequences and reference name

`df.loc[keep_samples, :].to_csv(output_metadata, sep='\t', index=True)`: failed by below error 

>KeyError: "Passing list-likes to .loc or [] with any missing labels is no longer supported. The following labels were missing: Index(['WIN-AH-2021-OTH-0038-5-M1571316', 'MN908947.3',\n       
'WIN-AH-2021-OTH-0038-2-M1571295'],\n      dtype='object', name='Virus_name'). See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike"

peterk87 commented:
> `/bin/filter_msa.py`:
> 
> `df = pd.read_table(input_metadata, index_col=0)`: dataframe of input_metadata does not contain strains of input sequences and reference sequence
> 
> `df.loc[keep_samples, :].to_csv(output_metadata, sep='\t', index=True)`: failed by below error
> 
> > KeyError: "Passing list-likes to .loc or [] with any missing labels is no longer supported. The following labels were missing: Index(['WIN-AH-2021-OTH-0038-5-M1571316', 'MN908947.3',\n
> > 'WIN-AH-2021-OTH-0038-2-M1571295'],\n      dtype='object', name='Virus_name'). See https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike"

Thanks @nhhaidee 

Please be aware that this is PR is a draft and not ready to be reviewed quite yet.

I'll let you know when it's ready. There is more testing that I need to do on my end.

nhhaidee commented:
@peterk87 OK. I just also do some tests at my end, will report bug if found so that you can fix it soon

peterk87 committed:
WIP: changes after testing on full GISAID dataset
refactored filter_gisaid.py
  added `--max-gisaid-seqs` to limit number of GISAID sequences to pull from large lineages
  changed stats output to JSON for easier programmatic uses
fixed index selection issue in prune_tree.py and filter_msa.py
combining Pangolin results and GISAID metadata into the same fields
Nextalign now using more cores and output files now explicit
removed unused process nw_reroot.nf
fixed issue with `--select_gisaid_metadata` workflow param due to default value set as "''"
added workflow params to skip Nextclade and R SNP tree processes

peterk87 committed:
generate software_version.tsv from scraped versions

peterk87 committed:
fix ci.yml

peterk87 committed:
try fix ci.yml

peterk87 committed:
fix markdownlint issues

peterk87 committed:
add artifact upload steps to ci.yml

peterk87 committed:
add iqtree log artifact upload step on failure to ci.yml

peterk87 committed:
more descriptive workflow params; updated nextflow_schema.json

nhhaidee commented on bin/filter_msa.py:
```diff
@@ -0,0 +1,128 @@
+    for sample, seq in sample_seq.items():
+        seq_samples[seq].add(sample)
+    return seq_samples
+
```
A Python set containing hashes of seq str and values are sequence trains, as I know there are some Canadian strains which have same sequences str with other countries, so just a concern that we may miss Canadian Sequences while keep sequences from other countries?

peterk87 replied:
All sequence IDs from the "focus" country could be added to the `keep_samples` set so they would be output no matter what.

nhhaidee commented:
Hi @peterk87 ,

I have some minor comments as below:

- Code is now very clean and clear.
- Missing Credits: nhhaidee/scovtree was originally written by Hai Nguyen and refactored by Peter Kruczkiewicz.
  We are using scripts bin/align2alles.py, phylogenetic_tree_snps.r of ncov-tools (https://github.com/jts/ncov-tools) for plotting phylogenetic tree with SNP, please credit them.
- `bin/filter_msa.py`: please see my concern in comments for line of code
- Unable to run workflow with conda profile:
>(nextflowenv) hnguyen@hnguyen-Precision-7750:~/Downloads/scovtree-refactor_1/scovtree-refactor$ nextflow run /home/hnguyen/Downloads/scovtree-refactor_1/scovtree-refactor -profile test,conda
N E X T F L O W  ~  version 21.04.0
Launching `/home/hnguyen/Downloads/scovtree-refactor_1/scovtree-refactor/main.nf` [high_almeida] - revision: 6216c3244d
Unknown method invocation `isEmpty` on Boolean type

Thanks,
Hai

nhhaidee commented on bin/filter_gisaid.py:
```diff
@@ -0,0 +1,252 @@
+            logging.info(f'Could not find "N_Content" GISAID metadata field. '
+                         f'Using equal probability weights for down-sampling')
+            weights = None
+        sampled_gisaid = df_subset.index.to_series()
```
Failed when with test mixing consensus sequences (B.1.2 + B.1.617.2 + B.1.618)
[Error_Lineages_B1_2_B1_618_B1_617_2.txt](https://github.com/nhhaidee/scovtree/files/6711516/Error_Lineages_B1_2_B1_618_B1_617_2.txt)

I sent you sequences to reproduce error via email

peterk87 replied:
Okay sounds good. I'd be interested to see the entire traceback for the script (`.command.log` file for the process) to see if there are more clues there.

nhhaidee replied:
[command.log](https://github.com/nhhaidee/scovtree/files/6711934/command.log)

peterk87 replied:
I'm having trouble recreating this error on my end. 

I noticed this error message in your `.command.log` file:

```
│ [Errno 2] No such file or directory: '/home/hnguyen/Downloads/scovtree-refac │
│ tor/work/7c/eb457e14d5781e2deb6a78b9e7d38e/mtrand.pyx'
```

I've added a `try/except` to catch `ValueError` exceptions when sampling with weights in 3bcc1e1a8c07230c4ec16f502e22f890cbc9d2c4

nhhaidee replied:
I just tried with updated code, the same problem happen and I can see work directory created. Could you try with data I sent you via email at your end?

nhhaidee replied:
Oh! I just updated your latest code and it seems to be OK now. The workflow is still running and will let you know when it is done

nhhaidee replied:
Test is passed so mark as Resolved

peterk87 committed:
update README.md
update ggtree to 3.0.1
update pangolin to 3.1.4
update nextclade to 0.14.4
update mafft to 7.480
update iqtree to 2.1.4_beta
add process cpu and mem reqs for certain processes in conf/base.config

peterk87 commented:
Thanks @nhhaidee  for taking a look at the changes I made. Some things like #29 and #19 aren't implemented here yet.

>     * Code is now very clean and clear.

Thanks!
 
>     * Missing Credits: nhhaidee/scovtree was originally written by Hai Nguyen and refactored by Peter Kruczkiewicz.
>       We are using scripts bin/align2alles.py, phylogenetic_tree_snps.r of ncov-tools (https://github.com/jts/ncov-tools) for plotting phylogenetic tree with SNP, please credit them.

Added in 9c75ffb

>     * Unable to run workflow with conda profile:

Do you have Conda installed and accessible on your `PATH`? I think the `base` env needs to be activated too.

I'm not sure what commit you're using. I don't think 6216c3244d corresponds to any commit I've made. Did you `git clone` the repo?

nhhaidee commented:
The code is now able to support multiple lineages, please attached for the test result for lineages are B.1.618 and B1.617.2 
[shiptv_B1_617_2_B1_618.zip](https://github.com/nhhaidee/scovtree/files/6712000/shiptv_B1_617_2_B1_618.zip)

There are 2 concerns that I have:

- Downsampling multiple sequence alignment by  `bin/filter_msa.py` may drop or keep low number of strains for low abundance lineages (Ex: < 200 sequences like B.1.618). Below is the code that I did test, keep entire low abundance lineages and sampling large abundance lineages only, please let me know your comment on this
```python
            df_total_len = len(df_less_n_gaps)
            df_stat = df_less_n_gaps.groupby('lineage').count()
            # Calculate abundance of lineages
            df_stat['abundances'] = df_stat['strain'] / df_total_len
            # For this case < 10% is considered as low abundance
            df_low_abundance_stat = df_stat[df_stat['abundances'] < 0.1]
            lineages_low_abundance = df_low_abundance_stat.index.tolist()
            df_low_abundance = df_less_n_gaps[df_less_n_gaps['lineage'].isin(lineages_low_abundance)]
            df_less_n_gaps.drop(df_less_n_gaps.index[df_less_n_gaps['lineage'].isin(lineages_low_abundance)],
                                inplace=True)
            df_sample = df_less_n_gaps.groupby('lineage').apply(
                lambda x: x.sample(int((len(x) / df_total_len) * threshold))).reset_index(drop=True)

            df_sample_strains = pd.concat([df_low_abundance, df_sample])

```
-  Prune tree for shiptv: how do we set criteria for branch distance? (https://github.com/nhhaidee/scovtree/issues/29). You can look at the tree attached and pay attention to the subtree of lineage B.1.618 with input sequence is WIN-AH-201-OTH-0041, I set max_tax = 75, the subtree is still quite large, if set max_tax is small (Ex: 20) then the tree will be too small. It would be nice if we could cut down unnecessary taxa, keep the tree enough for reporting

How do you think if we have params are `max_branch_distance` and `max_taxa` to make tree enough for reporting?
```python
max_branch_distance = 0.0005
def get_neighbors(tree: Tree, user_taxa: Set[str], clade_neighbors: Set[str], max_taxa: int) -> Set[str]:
    # TODO: implement adding neighboring taxa equally for all user taxa based on branch distance [peterk87 2021-06-22]
    for user_taxon in user_taxa:
        user_taxon_node = list(tree.find_elements(target=user_taxon))[0]
        clade: Clade
        for clade in tree.get_path(user_taxon_node)[::-1]:
            n_terminals = clade.count_terminals()
            if n_terminals <= max_taxa:
                clade_neighbors |= {nc.name for nc in clade.get_terminals() if tree.distance(user_taxon_node,nc) <= max_branch_distance}
            else:
                break
    return clade_neighbors

```

nhhaidee commented on bin/prune_tree.py:
```diff
@@ -0,0 +1,77 @@
+            else:
+                break
+    return clade_neighbors
+
```
I think that lines of code fixed issue https://github.com/nhhaidee/scovtree/issues/19

nhhaidee commented on bin/prune_tree.py:
```diff
@@ -0,0 +1,77 @@
+
+def get_neighbors(tree: Tree, user_taxa: Set[str], clade_neighbors: Set[str], max_taxa: int) -> Set[str]:
+    # TODO: implement adding neighboring taxa equally for all user taxa based on branch distance [peterk87 2021-06-22]
+    for user_taxon in user_taxa:
```
I think that lines of code fixed issue https://github.com/nhhaidee/scovtree/issues/19

peterk87 replied:
I'll have to do more testing, but from what I've seen so far, the first user taxon in the list gets all the neighboring taxa. #29 needs to be implemented so that the resulting tree is more balanced.

nhhaidee replied:
Marked as Resolve, the issue will be discussed in another PR

nhhaidee commented:
>* Unable to run workflow with conda profile:

> Do you have Conda installed and accessible on your PATH? I think the base env needs to be activated too.
> I'm not sure what commit you're using. I don't think 6216c3244d corresponds to any commit I've made. Did you git clone the repo?

Work now at my end

nhhaidee commented on modules/local/nextalign.nf:
```diff
@@ -0,0 +1,42 @@
+      saveAs: { filename -> saveFiles(filename:filename, options:params.options, publish_dir:getSoftwareName(task.process), meta:[:], publish_by_meta:[]) }
+
+  conda (params.enable_conda ? 'nextalign==0.2.0--h9ee0642_1' : null)
+  if (workflow.containerEngine == 'singularity' && !params.singularity_pull_docker_container) {
```
There is  a bug when creating conda env for nextalign, please update as below
`conda (params.enable_conda ? 'bioconda::nextalign==0.2.0' : null)`

peterk87 replied:
Fixed in ceeb18fb6eb19ffad2a282272e15d611fcb0bcf7

nhhaidee commented on bin/merge_metadata.py:
```diff
@@ -1,64 +1,65 @@
+        logging.info(f'Selecting specified metadata fields/columns from '
+                     f'"{metadata_input}": {select_metadata_fields}')
+        df_metadata = select_fields(df_metadata, select_metadata_fields)
+    dfs = [df_metadata]
```
KeyError: 'Pango_lineage' when `select_gisaid_metadata` param does not contain `Pango_lineage` field. I ran test `nextflow run /home/hnguyen/Documents/nextflow/scovtree-refactor/scovtree/ -profile test_gisaid_custom,docker`. 

Please log file attached 
[select_metadata_gisaid_command.log](https://github.com/nhhaidee/scovtree/files/6716315/select_metadata_gisaid_command.log)

nhhaidee replied:
The error also happens when `Pangolin_version` missed so these columns dropped.

peterk87 replied:
Fixed in ceeb18fb6eb19ffad2a282272e15d611fcb0bcf7

nhhaidee commented on conf/test_gisaid.config:
```diff
@@ -21,22 +21,14 @@ params {
   xambig               = 3000
-  msa_threshold_filter = 10000
+  msa_threshold_filter = 100
   max_taxa             = 50
```
Can you update these params as in nextflow.conf?
```
  gisaid_sequences
  gisaid_metadata 
  gisaid_country 
  gisaid_region 
  gisaid_min_length 
  gisaid_max_length
  gisaid_max_ambig 
  max_gisaid_filtered_seqs  
```

peterk87 replied:
Fixed in ceeb18fb6eb19ffad2a282272e15d611fcb0bcf7

nhhaidee commented on conf/test_gisaid_custom.config:
```diff
@@ -22,10 +22,8 @@ params {
-
   country              = ''
   region               = ''
   sample_lineage       = ''
```
Can you update these params as in nextflow.conf?
```
  gisaid_sequences
  gisaid_metadata 
  gisaid_country 
  gisaid_region 
  gisaid_min_length 
  gisaid_max_length
  gisaid_max_ambig 
  max_gisaid_filtered_seqs
```

peterk87 replied:
Fixed in ceeb18fb6eb19ffad2a282272e15d611fcb0bcf7

peterk87 committed:
merge_metadata.py: check that 'Pango_lineage' and 'Pangolin_version' columns exist before trying combine_first
base.config: fix processes not being selected by name; using labels to select processes
updated test configs
nextalign.nf: specifying nextalign from bioconda channel when using conda

peterk87 commented:
> The code is now able to support multiple lineages, please attached for the test result for lineages are B.1.618 and B1.617.2
> [shiptv_B1_617_2_B1_618.zip](https://github.com/nhhaidee/scovtree/files/6712000/shiptv_B1_617_2_B1_618.zip)

@nhhaidee Are you talking about the changes in this PR? This PR is just for refactoring and restructuring. I didn't really touch issue #19 or #29 

> There are 2 concerns that I have:
> 
>     * Downsampling multiple sequence alignment by  `bin/filter_msa.py` may drop or keep low number of strains for low abundance lineages (Ex: < 200 sequences like B.1.618). Below is the code that I did test, keep entire low abundance lineages and sampling large abundance lineages only, please let me know your comment on this
> 
> 
> ```python
>             df_total_len = len(df_less_n_gaps)
>             df_stat = df_less_n_gaps.groupby('lineage').count()
>             # Calculate abundance of lineages
>             df_stat['abundances'] = df_stat['strain'] / df_total_len
>             # For this case < 10% is considered as low abundance
>             df_low_abundance_stat = df_stat[df_stat['abundances'] < 0.1]
>             lineages_low_abundance = df_low_abundance_stat.index.tolist()
>             df_low_abundance = df_less_n_gaps[df_less_n_gaps['lineage'].isin(lineages_low_abundance)]
>             df_less_n_gaps.drop(df_less_n_gaps.index[df_less_n_gaps['lineage'].isin(lineages_low_abundance)],
>                                 inplace=True)
>             df_sample = df_less_n_gaps.groupby('lineage').apply(
>                 lambda x: x.sample(int((len(x) / df_total_len) * threshold))).reset_index(drop=True)
> 
>             df_sample_strains = pd.concat([df_low_abundance, df_sample])
> ```
> 
>     * Prune tree for shiptv: how do we set criteria for branch distance? ([Add neighbouring taxa using round-robin algorithm #29](https://github.com/nhhaidee/scovtree/issues/29)). You can look at the tree attached and pay attention to the subtree of lineage B.1.618 with input sequence is WIN-AH-201-OTH-0041, I set max_tax = 75, the subtree is still quite large, if set max_tax is small (Ex: 20) then the tree will be too small. It would be nice if we could cut down unnecessary taxa, keep the tree enough for reporting
> 
> 
> How do you think if we have params are `max_branch_distance` and `max_taxa` to make tree enough for reporting?
> 
> ```python
> max_branch_distance = 0.0005
> def get_neighbors(tree: Tree, user_taxa: Set[str], clade_neighbors: Set[str], max_taxa: int) -> Set[str]:
>     # TODO: implement adding neighboring taxa equally for all user taxa based on branch distance [peterk87 2021-06-22]
>     for user_taxon in user_taxa:
>         user_taxon_node = list(tree.find_elements(target=user_taxon))[0]
>         clade: Clade
>         for clade in tree.get_path(user_taxon_node)[::-1]:
>             n_terminals = clade.count_terminals()
>             if n_terminals <= max_taxa:
>                 clade_neighbors |= {nc.name for nc in clade.get_terminals() if tree.distance(user_taxon_node,nc) <= max_branch_distance}
>             else:
>                 break
>     return clade_neighbors
> ```

Let's move these to another PR. 

This PR is mostly to refactor and restructure the workflow and code so that further changes can be implemented more easily.

nhhaidee commented:
OK @peterk87 For issue https://github.com/nhhaidee/scovtree/issues/29 and https://github.com/nhhaidee/scovtree/issues/19, will be discussed in another PR. I will wait for your final fixes, and run test at my end as well, if they are OK then we will merge and close this PR

peterk87 committed:
filter_msa.py: extracted `keep_seqs_from_country` function to add seqs from focus country to `keep_samples` set
removed markdown_to_html.py; unused
updated usage.md
added convenience script nextflow_schema_json_to_markdown.py to create Markdown for workflow params from nextflow_schema.json file

peterk87 committed:
prune_tree.py: added leaf list construction by iteratively adding nearest neighboring taxon for each user taxon until max taxa for pruned tree hit (addresses issues #19 and #29)
updated output.md
updated pangolin to 3.1.5
added workflow onComplete and onError messages

peterk87 committed:
filter_gisaid.py: added try/except to catch ValueError exception when sampling with N_Content weights; sampling will proceed with no weights (equal weighting)

peterk87 commented:
@nhhaidee I added some code in aea78817293eb3489100b83fbf3205916760ee15 to `bin/prune_tree.py` to address #19 and #29. Trees produced should hopefully be a bit more balanced in terms of neighboring taxa for each user taxon.

peterk87 committed:
fix markdown lint issues

nhhaidee commented:
>@nhhaidee I added some code in aea7881 to bin/prune_tree.py to address #19 and #29. Trees produced should hopefully be a bit more balanced in terms of neighboring taxa for each user taxon.

I will test with my data, I have tested with below code and tree seems to be balanced

```python
def get_neighbors(tree: Tree, user_taxa: Set[str], clade_neighbors: Set[str], max_taxa: int) -> Set[str]:
    # TODO: implement adding neighboring taxa equally for all user taxa based on branch distance [peterk87 2021-06-22]
    for user_taxon in user_taxa:
        user_taxon_node = list(tree.find_elements(target=user_taxon))[0]
        clade: Clade
        for clade in tree.get_path(user_taxon_node)[::-1]:
            n_terminals = clade.count_terminals()
            if n_terminals <= max_taxa * 5: # condition to make sure tree is not too small or too large which is run very slow
                clade_neighbors_user_taxon = {}
                for nc in clade.get_terminals():
                    clade_neighbors_user_taxon[nc.name] = tree.distance(nc.name,user_taxon_node)
                sorted_taxa_neighbors  = dict(sorted(clade_neighbors_user_taxon.items(), key=lambda item: item[1]))
                clade_neighbors |= {name for i, name in enumerate(sorted_taxa_neighbors.keys()) if i< max_taxa}
    return clade_neighbors
```

peterk87 commented:
@nhhaidee 

I have a few issues with the above code:

- too much nesting making it hard to read and signaling that the implementation is inefficient
- the same `tree.distance` calculation may be repeated multiple times. It's also a very slow calculation so repeating it is going to significantly increase the runtime esp with a large number of user input sequences.
- `5` is a ["magic" number](https://en.wikipedia.org/wiki/Magic_number_%28programming%29). Magic numbers should be avoided whenever possible. I don't see why it's necessary.
- most importantly: the first user taxon will use up most or all the space in the leaf list. Please see my implementation where, for each user taxon, the nearest neighboring taxon is added until the desired number of taxa is reached https://github.com/nhhaidee/scovtree/blob/aea78817293eb3489100b83fbf3205916760ee15/bin/prune_tree.py#L80-L85

Please try your code and my code with `-profile test_gisaid` and the full GISAID data (also make sure you set `--max_cpus {>2} --max_memory {>6 GB}` to override test profile settings) to see the differences in the resulting trees.

nhhaidee commented:
OK @peterk87 I am just trying with my code and will test your code as well to see how the tree look like.

FYI... I am doing test for failure with my full data

nhhaidee commented:
@peterk87 The tree looks very nice with your code.
[shiptv_full_data.zip](https://github.com/nhhaidee/scovtree/files/6719115/shiptv_full_data.zip)

I will do testing with more test cases

nhhaidee commented:
Test cases:
Sample Lineages in GISAID  B.1.2 (> 100,000 sequences); B.1.617.2 ( > 37,000 sequences); B.1.618 ( > 180 sequences)
Input sequences: B.1.2 has 2 sequences, B.1.617.2 has 1 sequence, B.1.618 has 1 sequence

Test Case 1: 2 input consensus sequences belong to B.1.2 lineage (Passed)
Test Case 2: 1 input consensus sequence belong to B.1.617.2 lineage (No Run)
Test Case 3: 1 input consensus sequence belong to B.1.618 lineage (Passed)
Test Case 4: Mix consensus sequences (B.1.2 + B.1.617.2) (No Run)
Test Case 5: Mix consensus sequences (B.1.2 + B.1.618) (No Run)
Test Case 6: Mix consensus sequences (B.1.617.2 + B.1.618) (Failed)
Test Case 7: Mix consensus sequences (B.1.2 + B.1.617.2 + B.1.618) (Passed)
Test Case 8: test_gisaid, test_gisaid_custom (Passed)

Test Case 6 Failed with SHIPTV Process with error `RecursionError: maximum recursion depth exceeded while calling a Python object`. Please see log file 
[comman_shiptv.log](https://github.com/nhhaidee/scovtree/files/6719351/comman_shiptv.log)

I will double check test case 6 again.

nhhaidee commented:
Updated tests:

Test cases:
Sample Lineages in GISAID B.1.2 (> 100,000 sequences); B.1.617.2 ( > 37,000 sequences); B.1.618 ( > 180 sequences)
Input sequences: B.1.2 has 2 sequences, B.1.617.2 has 1 sequence, B.1.618 has 1 sequence

Test Case 1: 2 input consensus sequences belong to B.1.2 lineage (Passed)
Test Case 2: 1 input consensus sequence belong to B.1.617.2 lineage (Passed)
Test Case 3: 1 input consensus sequence belong to B.1.618 lineage (Passed)
Test Case 4: Mix consensus sequences (B.1.2 + B.1.617.2) (Failed)
Test Case 5: Mix consensus sequences (B.1.2 + B.1.618) (Passed)
Test Case 6: Mix consensus sequences (B.1.617.2 + B.1.618) (Failed)
Test Case 7: Mix consensus sequences (B.1.2 + B.1.617.2 + B.1.618) (Passed)
Test Case 8: test_gisaid, test_gisaid_custom (Passed)

Failures related to lineage B.1.617.2. There is no problem with SHIPTV process with commit (https://github.com/peterk87/scovtree/commit/97ec512dc1fa59ced37065d1fc006c9bdfa90508). It seems that it is failed after changing code to handle issue https://github.com/nhhaidee/scovtree/issues/29 and https://github.com/nhhaidee/scovtree/issues/19

peterk87 committed:
shiptv.nf: remove shiptv output newick due to BioPython Phylo recursion issues

peterk87 commented:
> Test Case 6 Failed with SHIPTV Process with error `RecursionError: maximum recursion depth exceeded while calling a Python object`. Please see log file
> [comman_shiptv.log](https://github.com/nhhaidee/scovtree/files/6719351/comman_shiptv.log)
> 
> I will double check test case 6 again.

It seems to be an issue with BioPython's Phylo when writing a large tree. I've removed shiptv output of a Newick tree in e905072688eba8d489a52f262d54da4f9e600c25 and created [an issue for shiptv](https://github.com/peterk87/shiptv/issues/17).

nhhaidee commented:
TC4 and TC6 PASSED. So I think there is no problem with GISIAD Phylolegenetic workflow

`--skip_nextclade` param does not work with phylogenetic analysis workflows (without filtering GISAID)
[skip_nextclade_does_not_work.log](https://github.com/nhhaidee/scovtree/files/6728523/skip_nextclade_does_not_work.log)

Do you want me to merge and close this PR and open new PR for above issue?

peterk87 committed:
shiptv_metadata.py: convert aa_mutation_matrix, leaflist, metadata_output from CLI arguments to options; add check if aa_mutation_matrix provided as input before merging
shiptv_metadata.nf: aa_mutation_matrix is now an optional argument to shiptv_metadata.py

peterk87 commented:
> TC4 and TC6 PASSED. So I think there is no problem with GISIAD Phylolegenetic workflow
> 
> `--skip_nextclade` param does not work with phylogenetic analysis workflows (without filtering GISAID)
> [skip_nextclade_does_not_work.log](https://github.com/nhhaidee/scovtree/files/6728523/skip_nextclade_does_not_work.log)
> 
> Do you want me to merge and close this PR and open new PR for above issue?

Nice catch @nhhaidee ! That's definitely a bug. It should be fixed in b44c93afd557b86bcddc3f73a44c9b91a74d7797 

I think the real issue is that there are 2 separate metadata merging scripts/processes that do very similar things. This functionality should be consolidated into a single script. I've created an issue (#31) for this.

nhhaidee commented:
It works now. I will do some testing and then I will merge and close this PR

nhhaidee commented:
All seems to be good now. I will merge/close this PR. The test is still failed intermittently because of IQTree, there is no problem when run the test locally.

nhhaidee commented:
Thank @peterk87 for great help on this.
